### PR TITLE
Add sLDA for Gensim

### PR DESCRIPTION
**Pull request by souravsingh**
I am making an initial PR for the project to get reviews from reviewers and getting fixes made.

This is my first big PR to Gensim.

### CODE REVIEW

**Comment by souravsingh** [comment]
PEP8 failures in code. Should be fixed soon.

File: gensim/models/sldamodel.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: A lot of PEP8 issues, please look [here](https://travis-ci.org/RaRe-Technologies/gensim/jobs/282716651#L1018).

Comment (menshikh-iv) [reply gensim/models/sldamodel.py]: + Missed `sldamodel.rst` in `docs/src/models`

File: gensim/models/sldamodel.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+from scipy.special import gammaln, psi  # gamma function utils
+from scipy.special import polygamma
+from six.moves import xrange
+
+from gensim import interfaces, utils, matutils
+from gensim.matutils import dirichlet_expectation
+from gensim.models import basemodel, CoherenceModel
+
+# log(sum(exp(x))) that tries to avoid overflow
+try:
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: Useless try/catch, for current minimal required SciPy version `from scipy.misc import logsumexp` is enough.

File: gensim/models/sldamodel.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+#
+# Copyright (C) 2011 Radim Rehurek <radimrehurek@seznam.cz>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+import logging
+import numpy as np
+import six
+from scipy.special import gammaln, psi  # gamma function utils
+from scipy.special import polygamma
+from six.moves import xrange
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: `xrange` is never used, only `range`, replace it.

File: gensim/models/sldamodel.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2011 Radim Rehurek <radimrehurek@seznam.cz>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+import logging
+import numpy as np
+import six
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: No need to import all `six`, you need only `string_types`.

File: gensim/models/sldamodel.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    # try importing from here if older scipy is installed
+    from scipy.maxentropy import logsumexp
+except ImportError:
+    # maxentropy has been removed in recent releases, logsumexp now in misc
+    from scipy.misc import logsumexp
+
+
+logger = logging.getLogger('gensim.models.sldamodel')
+
+def dirichlet_expectation(alpha):
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: useless parentheses (here and anywhere)

File: gensim/models/sldamodel.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    """
+    For a vector theta ~ Dir(alpha), computes E[log(theta)] given alpha.
+    """
+    if (len(alpha.shape) == 1):
+        return(psi(alpha) - psi(n.sum(alpha)))
+    return(psi(alpha) - psi(n.sum(alpha, 1))[:, n.newaxis])
+
+def get_lambda(self):
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: Standart - 4 spaces for one indentation (not 8).

File: gensim/models/sldamodel.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        return self.eta + self.sstats
+
+def get_Elogbeta(self):
+        return dirichlet_expectation(self.get_lambda())
+
+class sLdaModel(interfaces.TransformationABC, basemodel.BaseTopicModel):
+
+    def __init__(self, corpus=None, n_topics=100, alpha='symmetric', beta, mu, nu, nu2,
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: Syntax error: args after kwargs.
We already discussed that you need to run and check your code first (before PR), why you continue to push broken code?

File: gensim/models/sldamodel.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            Mean of regression coefficients (eta).
+        nu2 : float
+            Variance of regression coefficients (eta).
+        sigma2 : float
+            Variance of response (y).
+        n_iter : int, default=100
+            Number of iterations of Gibbs sampler
+        n_report_iter : int, default=10
+            Number of iterations of Gibbs sampler between progress reports.
+        seed : int, optional
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: Missed parameters: report_iter, iterations, nu, corpus
New parameters (that not defined in \_\_init\_\_): n_iter, n_report_iter.

File: gensim/models/sldamodel.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+            shown.append((i, topic_))
+            if log:
+                logger.info("topic #%i (%.3f): %s", i, self.alpha[i], topic_)
+
+        return shown
+
+    def show_topic(self, topicid, topn=10):
+        """
+        Args:
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: missed `topn` param

File: gensim/models/sldamodel.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+        
+    def accuracy(self, goldlabel):
+        right = 0
+        for d in range(0, self._D):
+            if (self._predictions[d] == goldlabel[d]):
+                right = right + 1
+        accuracy = float(right) / float(self._D)
+        return accuracy
+    
+    def save_parameters(self):
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: Why this function? We already have `save`?

File: gensim/test/test_sldamodel.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2010 Radim Rehurek <radimrehurek@seznam.cz>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+"""
+Automated tests for checking transformation algorithms (the models package).
```
Comment (menshikh-iv) [inline gensim/test/test_sldamodel.py]: Really?

File: gensim/test/test_sldamodel.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+"""
+
+
+import logging
```
Comment (menshikh-iv) [inline gensim/test/test_sldamodel.py]: A lot of useless imports

File: gensim/test/test_sldamodel.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+import os.path
+import tempfile
+import numbers
+
+import six
+import numpy as np
+import scipy.linalg
+
+from gensim.corpora import mmcorpus, Dictionary
+from gensim.models import ldamodel, ldamulticore
```
Comment (menshikh-iv) [inline gensim/test/test_sldamodel.py]: What for? useless too.

File: gensim/models/sldamodel.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+    def __init__(self, corpus=None, n_topics=100, alpha='symmetric', beta, mu, nu, nu2,
+                 sigma2, iterations = 100, report_iter=10, seed=None):
+    """
+    Supervised latent Dirichlet allocation, using collapsed Gibbs
+    sampling.
+    Args
+        n_topics : int
+            Number of topics
+        alpha : array-like, shape = (n_topics,)
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: but I see `symmetric` param, it's not `array-like`, it's string

File: gensim/models/sldamodel.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+            (10 words for top 10 topics, by default).
+        """
+        if num_topics < 0 or num_topics >= self.num_topics:
+            num_topics = self.num_topics
+            chosen_topics = range(num_topics)
+        else:
+            num_topics = min(num_topics, self.num_topics)
+
+            # add a little random jitter, to randomize results around the same alpha
+            sort_alpha = self.alpha + 0.0001 * self.random_state.rand(len(self.alpha))
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: Concatenation string with float.

File: gensim/models/sldamodel.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+        return(psi(alpha) - psi(n.sum(alpha)))
+    return(psi(alpha) - psi(n.sum(alpha, 1))[:, n.newaxis])
+
+def get_lambda(self):
+        return self.eta + self.sstats
+
+def get_Elogbeta(self):
+        return dirichlet_expectation(self.get_lambda())
+
+class sLdaModel(interfaces.TransformationABC, basemodel.BaseTopicModel):
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: SLdaModel, because CamelCase

**Comment by menshikh-iv** [comment]
Attention @sanikak96 @devshreesk @sakshi-s-jain, if I'll see no contribution from you for before next meeting and you'll continue miss hangouts, I'll finish working with your team.

**Comment by souravsingh** [comment]
@menshikh-iv Thanks for the review on the PR. I will quickly fix all the issues mentioned.

File: gensim/models/sldamodel.py [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+    dprior = -(gradf - b) / q
+
+    if all(rho * dprior + prior > 0):
+        prior += rho * dprior
+    else:
+        logger.warning("updated prior not positive")
+
+    return prior
+
+class sLdaState(utils.SaveLoad):
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: Same as `LdaState`, use `LdaState`

File: gensim/models/sldamodel.py [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+    
+logger = logging.getLogger('gensim.models.sldamodel')
+
+def update_dir_prior(prior, N, logphat, rho):
+    """
+    Updates a given prior using Newton's method, described in
+    **Huang: Maximum Likelihood Estimation of Dirichlet Distribution Parameters.**
+    http://jonathan-huang.org/research/dirichlet/dirichlet.pdf
+    """
+    dprior = np.copy(prior)  # TODO: unused var???
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: ??

File: gensim/models/sldamodel.py [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+try:
+    from scipy.special import logsumexp
+except ImportError:
+    from scipy.misc import logsumexp
+    
+logger = logging.getLogger('gensim.models.sldamodel')
+
+def update_dir_prior(prior, N, logphat, rho):
+    """
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: please use [numpy-style](https://github.com/RaRe-Technologies/gensim/wiki/Developer-page#docstrings) docstrings (here and everywhere).

File: gensim/models/sldamodel.py [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+        self.sstats += other.sstats
+        self.numdocs = targetsize
+
+    def get_lambda(self):
+        return self.eta + self.sstats
+
+    def get_Elogbeta(self):
+        return dirichlet_expectation(self.get_lambda())
+
+class sLdaModel(interfaces.TransformationABC, basemodel.BaseTopicModel):
```
Comment (menshikh-iv) [inline gensim/models/sldamodel.py]: `CamelCase`: `sLdaModel` -> `SLdaModel`

File: gensim/test/test_sldamodel.py [hunk 21]
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python
```
Comment (menshikh-iv) [inline gensim/test/test_sldamodel.py]: It's time to do tests

File: gensim/models/sldamodel.py (hunk 18 above)
Comment (souravsingh) [reply gensim/models/sldamodel.py]: I had a feeling that there is an unused parameter for the ordering, so I put in the comment. I can remove it if needed.

File: gensim/models/sldamodel.py (hunk 19 above)
Comment (souravsingh) [reply gensim/models/sldamodel.py]: Roger

File: gensim/models/sldamodel.py (hunk 20 above)
Comment (souravsingh) [reply gensim/models/sldamodel.py]: Ok

**Pull request closed without merging**